File: q1249/q1249.java
class Solution {
    public String minRemoveToMakeValid(String s) {
        ArrayDeque<Integer> stack = new ArrayDeque<>();
        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            // If it is a closing parenthesis, and there is a matching open one: remove both
            if (c == ')' && (!stack.isEmpty() && s.charAt(stack.peek()) == '(')) {
                stack.pop();    
            } 
            // Otherwise simply push it on to the stack for future tracking
            else if (c == ')' || c == '(') {
                stack.push(i);
            }
        }
        
        // Rebuild the string by removing the parentheses that are on the stack (not properly closed)
        StringBuilder sb = new StringBuilder(s);
        while (!stack.isEmpty()) {
            int i = stack.pop();
            sb.deleteCharAt(i);
        }
        return sb.toString();
    }
}

File: q1249/q1249.py
class Solution:
    def minRemoveToMakeValid(self, s: str) -> str:
        stack = []
        for i in range(len(s)):
            c = s[i]
            if c == ')' and (stack and s[stack[-1]] == '('): stack.pop()
            elif (c == '(' or c == ')'): stack.append(i)
        
        res = []
        for i in range(len(s)):
            if stack and stack[0] == i: 
                stack.pop(0)
                continue
            res.append(s[i])
        return "".join(res)